File: uc-deliveries/static/app/models/preview.mjs
const UNASSIGNED_DELIVERY_HARD_PENALTY = 1_000_000;

export function buildPreview(plan) {
  const assigned = new Set();
  const vehicles = [];
  const deliveries = (plan.deliveries || []).map((delivery) => ({
    deliveryId: delivery.id,
    label: delivery.label,
    kind: delivery.kind,
    demand: delivery.demand,
    minStartTime: delivery.minStartTime,
    maxEndTime: delivery.maxEndTime,
    serviceDuration: delivery.serviceDuration,
    assignedVehicleId: null,
    assignedVehicleName: null,
    sequence: null,
    arrivalTime: null,
    serviceStartTime: null,
    departureTime: null,
    lateSeconds: null,
  }));

  let capacityOverage = 0;
  let lateSeconds = 0;
  let travelSeconds = 0;

  for (const vehicle of plan.vehicles || []) {
    const metrics = computeVehicleMetrics(plan, vehicle);
    capacityOverage += metrics.capacityOverage;
    lateSeconds += metrics.totalLateSeconds;
    travelSeconds += metrics.totalTravelSeconds;
    vehicles.push(metrics);

    for (const stop of metrics.stops) {
      assigned.add(stop.deliveryId);
      const delivery = deliveries[stop.deliveryId];
      delivery.assignedVehicleId = vehicle.id;
      delivery.assignedVehicleName = vehicle.name;
      delivery.sequence = stop.sequence;
      delivery.arrivalTime = stop.arrivalTime;
      delivery.serviceStartTime = stop.serviceStartTime;
      delivery.departureTime = stop.departureTime;
      delivery.lateSeconds = stop.lateSeconds;
    }
  }

  const unassignedDeliveryIds = deliveries
    .filter((delivery) => !assigned.has(delivery.deliveryId))
    .map((delivery) => delivery.deliveryId);

  return {
    hardScore: -(
      unassignedDeliveryIds.length * UNASSIGNED_DELIVERY_HARD_PENALTY +
      capacityOverage +
      lateSeconds
    ),
    softScore: -travelSeconds,
    unassignedDeliveryIds,
    vehicles,
    deliveries,
  };
}

function computeVehicleMetrics(plan, vehicle) {
  const stops = [];
  let totalDemand = 0;
  let totalTravelSeconds = 0;
  let totalWaitSeconds = 0;
  let totalServiceSeconds = 0;
  let totalLateSeconds = 0;
  let endTime = vehicle.departureTime || 0;
  let currentTime = vehicle.departureTime || 0;

  for (const [sequence, deliveryId] of (vehicle.deliveryOrder || []).entries()) {
    const delivery = plan.deliveries[deliveryId];
    if (!delivery) continue;

    totalDemand += Number(delivery.demand || 0);
    const travel = 0;
    totalTravelSeconds += travel;
    const arrivalTime = currentTime + travel;
    const serviceStartTime = Math.max(arrivalTime, delivery.minStartTime || 0);
    const waitSeconds = Math.max(0, serviceStartTime - arrivalTime);
    const departureTime = serviceStartTime + Number(delivery.serviceDuration || 0);
    const lateSeconds = Math.max(0, departureTime - Number(delivery.maxEndTime || 0));

    totalWaitSeconds += waitSeconds;
    totalServiceSeconds += Number(delivery.serviceDuration || 0);
    totalLateSeconds += lateSeconds;
    currentTime = departureTime;
    endTime = departureTime;

    stops.push({
      deliveryId,
      label: delivery.label,
      kind: delivery.kind,
      sequence,
      demand: delivery.demand,
      minStartTime: delivery.minStartTime,
      maxEndTime: delivery.maxEndTime,
      arrivalTime,
      serviceStartTime,
      departureTime,
      travelSecondsFromPrevious: travel,
      waitSeconds,
      lateSeconds,
    });
  }

  if (stops.length) {
    endTime = currentTime;
  }

  return {
    vehicleId: vehicle.id,
    vehicleName: vehicle.name,
    totalDemand,
    capacityOverage: Math.max(0, totalDemand - Number(vehicle.capacity || 0)),
    stopCount: stops.length,
    totalTravelSeconds,
    totalWaitSeconds,
    totalServiceSeconds,
    totalLateSeconds,
    startTime: vehicle.departureTime || 0,
    endTime,
    stops,
  };
}

File: uc-deliveries/src/domain/preview.rs
//! Browser preview structs embedded in `Plan.view_state`.
//!
//! These values are derived from the domain plan before transport. They let the
//! frontend render route summaries and timelines without duplicating every
//! scoring rule.

use serde::{Deserialize, Serialize};

/// Records the road-network routing contract in transport payloads.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RoutingMode {
    #[default]
    RoadNetwork,
}

/// Selects which timeline rail the browser shows.
#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TimelineView {
    #[default]
    ByVehicle,
    ByDelivery,
}

/// UI-only state that travels with the plan between browser and backend.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlanViewState {
    #[serde(default)]
    pub timeline_view: TimelineView,
    pub selected_vehicle_id: Option<usize>,
    pub selected_delivery_id: Option<usize>,
    #[serde(default)]
    pub preview: Option<PlanPreview>,
}

/// Aggregate route and score preview for the full plan.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlanPreview {
    pub hard_score: i64,
    pub soft_score: i64,
    pub unassigned_delivery_ids: Vec<usize>,
    pub vehicles: Vec<VehiclePreview>,
    pub deliveries: Vec<DeliveryPreview>,
}

/// Per-vehicle route summary used by cards, lists, and timelines.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VehiclePreview {
    pub vehicle_id: usize,
    pub vehicle_name: String,
    pub total_demand: i32,
    pub capacity_overage: i32,
    pub stop_count: usize,
    pub total_travel_seconds: i64,
    pub total_wait_seconds: i64,
    pub total_service_seconds: i64,
    pub total_late_seconds: i64,
    pub start_time: i64,
    pub end_time: i64,
    pub stops: Vec<VehiclePreviewStop>,
}

/// One delivery stop on a vehicle timeline.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VehiclePreviewStop {
    pub delivery_id: usize,
    pub label: String,
    pub kind: String,
    pub sequence: usize,
    pub demand: i32,
    pub min_start_time: i64,
    pub max_end_time: i64,
    pub arrival_time: i64,
    pub service_start_time: i64,
    pub departure_time: i64,
    pub travel_seconds_from_previous: i64,
    pub wait_seconds: i64,
    pub late_seconds: i64,
}

/// Per-delivery assignment summary used by data tables and timelines.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeliveryPreview {
    pub delivery_id: usize,
    pub label: String,
    pub kind: String,
    pub demand: i32,
    pub min_start_time: i64,
    pub max_end_time: i64,
    pub service_duration: i64,
    pub assigned_vehicle_id: Option<usize>,
    pub assigned_vehicle_name: Option<String>,
    pub sequence: Option<usize>,
    pub arrival_time: Option<i64>,
    pub service_start_time: Option<i64>,
    pub departure_time: Option<i64>,
    pub late_seconds: Option<i64>,
}
